Create Robotframework Automated Tests transfered from PySys

gligorisaev opened the pull request:
## Proposed changes

tedge_connect Testcases that are transfered 
- c8y_restart_bridge - Removed
- tedge_connect_test_negative - Transferred to 
- tedge_connect_test_positive - Transferred to
- tedge_connect_test_sm_services - Transferred to
- tedge_disconnect_test_sm_services - Transferred to


## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1256 

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

matthiasbeyer commented:
So I don't know much about that transfer from pytest to robot, but from what I see this is not yet integrated in the github actions, so that's one point that is missing...

PradeepKiruvale commented on tests/RobotFramework/tedge/call_tedge.robot:
```diff
@@ -0,0 +1,62 @@
+*** Keywords ***
+Open Connection And Log In
+   SSHLibrary.Open Connection     ${HOST}
+   SSHLibrary.Login               ${USERNAME}        ${PASSWORD}
```
Add a line here.

PradeepKiruvale commented on tests/RobotFramework/tedge/call_tedge_config_list.robot:
```diff
@@ -0,0 +1,148 @@
+Suite Teardown         SSHLibrary.Close All Connections
+
+*** Variables ***
+${HOST}           192.168.99.110    #Insert the IP address if the default should not be used
```
Does this mean that when an environment variable `HOST` is not set, then it will go with the default value `192.168.99.110`?.

PradeepKiruvale commented on tests/RobotFramework/tedge/call_tedge_config_list.robot:
```diff
@@ -0,0 +1,148 @@
+*** Keywords ***
+Open Connection And Log In
+   SSHLibrary.Open Connection     ${HOST}
+   SSHLibrary.Login               ${USERNAME}        ${PASSWORD}
```
Add a new line here.

PradeepKiruvale commented on tests/RobotFramework/tedge/call_tedge_config_list.robot:
```diff
@@ -0,0 +1,148 @@
+${USERNAME}       pi    #Insert the username if the default should not be used
+${PASSWORD}       thinedge    #Insert the password if the default should not be used
+
+*** Tasks ***
```
I feel also add a test for `device.id`, this is not settable. So, one can test this feature.
For example

`sudo tedge config set device.id test`

Error: failed to set the configuration key: device.id with value: test.

Caused by:
    Readonly setting: The device id is read from the device certificate and cannot be set directly.
    To set 'device.id' to some <id>, you can use `tedge cert create --device-id <id>`.

PradeepKiruvale commented on tests/RobotFramework/tedge_connect/tedge_connect_test.robot:
```diff
@@ -0,0 +1,49 @@
+    Should Contain    ${stdout}    tedge-agent service successfully stopped and disabled!    #Expected message
+    Should Contain    ${stdout}    tedge-mapper-c8y service successfully stopped and disabled!    #Expected message
+    Log    ${stdout}
+
```
May be similar test is required for `azure` as well?

rina23q replied:
Yeah, definitely we need `tedge_connect_test_positive_azure` and `tedge_connect_test_negative_azure`. But I know why we didn't have them in PySys. We don't have a common Azure IoT Hub which is usable for testing. 

So, @PradeepKiruvale, if we already have it, it's time to ask @gligorisaev to write those tests.

PradeepKiruvale commented on tests/RobotFramework/tedge_connect/tedge_connect_test.robot:
```diff
@@ -0,0 +1,49 @@
+Open Connection And Log In
+   
+    SSHLibrary.Open Connection     ${HOST}
+    SSHLibrary.Login               ${USERNAME}        ${PASSWORD}
```
Add a line here.

PradeepKiruvale reviewed (commented):
Looks good to me.

Added some minor comments.

Will these tests be part of the `CI` pipeline?

rina23q commented:
> So I don't know much about that transfer from pytest to robot, but from what I see this is not yet integrated in the github actions, so that's one point that is missing...

It's not yet decided how we are going to run them. After merge? Nightly? I'd prefer to have it nightly workflow and run these tests from there, but haven't reached to consensus yet.

matthiasbeyer commented:
> GitHub hosted runner cuts internet connection from time to time.

Why is this an issue for these tests?

> How?
> After merge? Nightly? 

Why not doing it like every other PR, in the PR workflow? That's what CI is all about, isn't it?

Having to fix mistakes from the day before just because the tests only run once a day is just too much overhead IMO!

rina23q commented:
>>GitHub hosted runner cuts internet connection from time to time.

>Why is this an issue for these tests?
We may get false-negative results because of connection refused!

> Why not doing it like every other PR, in the PR workflow? That's what CI is all about, isn't it?
This PR is very limited, but we will port more tests. Actually, the whole set of the existing Pysys tests (system tests and end-to-end tests) took 2 hours per one run. That's why I am considering nightly. 

Anyway, it is not the scope of this PR. I prefer to have another place to discussion, probably on GitHub Discussions later. I don't have so deep thoughts about it yet.

rina23q commented on tests/RobotFramework/tedge_connect/tedge_connect_test.robot:
```diff
@@ -0,0 +1,49 @@
+    Should Contain    ${stdout}    tedge-agent service successfully started and enabled!    #Expected message
+    Should Contain    ${stdout}    tedge-mapper-c8y service successfully started and enabled!    #Expected message
+    Log    ${stdout}
+tedge_disconnect_test_sm_services
```
Add one line break.

rina23q replied:
I mean one line break between line 37 and 38.

rina23q commented on tests/RobotFramework/tedge_connect/tedge_connect_test.robot:
```diff
@@ -0,0 +1,49 @@
+
+tedge_connect_test_negative
+    Execute Command    sudo tedge disconnect c8y    #Disonnecting from Cumulocity IoT
+    ${stdout}    ${stderr}=    Execute Command    sudo tedge connect c8y --test    return_stderr=True    #Testing the status of the connection
```
It's a question. What ` ${stdout}    ${stderr}=` part means? We will record both `stdout` and `stderr`? I'm just curious because the next two lines don't use `${stdout}`.

gligorisaev replied:
> It's a question. What ` ${stdout} ${stderr}=` part means? We will record both `stdout` and `stderr`? I'm just curious because the next two lines don't use `${stdout}`.


${stdout} is just name of a variable, helps me to remember where I hae used it, it can be anything

rina23q reviewed (commented):
`tedge` directory should be removed from this PR as it is the scope of different issue.

gligorisaev commented:
Requested changes are implemented @rina23q please Review/close this issue

rina23q commented:
`call_tedge_config_list.robot` file is also added, which is not described above. Is it intended to add this file in this PR?

gligorisaev commented:
> `call_tedge_config_list.robot` file is also added, which is not described above. Is it intended to add this file in this PR?

Corrected, should be part of this PR

matthiasbeyer commented:
Why was this merged? It is not even integrated in the CI pipelines yet! Or am I missing something?
